### PR TITLE
Downloads show breadcrumb of category

### PR DESCRIPTION
**Pull request by Samangie**
Replace the category name in the list of categories with a simple breadcrumb. Now all parent categories are displayed and not just the current category name. 

The root/main category and the previous category are shortened after 30 characters. The current category name after 60 characters. The middle categories are shortened with three dots. If you hover over the category names or dots, the full category name will be displayed.

### CODE REVIEW

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 1]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+
+            }
+
+            // Flip the array to get the grandparents first
+            $categories['middle'] = array_reverse($categories['middle']);
+
+            $content =
+                sprintf(
+                    $_ARRAYLANG['TXT_DOWNLOADS_CATEGORIES_OF_CATEGORY'],
+                '&bdquo;'.
```
Comment (MichaelRitter) [inline modules/Downloads/Controller/DownloadsManager.class.php]: Indention error

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 2]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+                    $this->getCategoryElement($categories['root']) .
+                    $this->getCategoryElement($categories['middle']) .
+                    $this->getCategoryElement($categories['previous']) .
+                    $this->getCategoryElement($categories['current'])
+                     . '&ldquo;'
```
Comment (MichaelRitter) [inline modules/Downloads/Controller/DownloadsManager.class.php]: Indention error

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 3]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+    /**
+     * Get the name of the child category and all names of the parent categories
+     *
+     * @param \Cx\Modules\Downloads\Controller\Category $child child category
+     * @param array $names previous category names for recursion
+     * @return array array with all names
+     */
+    protected function getParentCategoryNamesByChild(
+        \Cx\Modules\Downloads\Controller\Category $child,
+        array $names = array()
```
Comment (MichaelRitter) [inline modules/Downloads/Controller/DownloadsManager.class.php]: Pass by reference to shorten memory usage.

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 1 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: done

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 2 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: done

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 3 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: done

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 4]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+                $text .= '...';
+            }
+
+        } else {
+            foreach ($category as $middle) {
+                $title .= '...' .$middle . "\n";
+            }
         }
+
+        return '<span title="'.$title.'">' . $text
```
Comment (daeppen) [inline modules/Downloads/Controller/DownloadsManager.class.php]: please do not add hard-coded html code. instead create html output using
`new \Cx\Core\Html\Model\Entity\HtmlElement('span');`
see https://wiki.cloudrexx.com/Development_Core_Html#Manual_Views

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 5]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+     *
+     * @param \Cx\Modules\Downloads\Controller\Category $child child category
+     * @param array $names previous category names for recursion
+     */
+    protected function getParentCategoryNamesByChild(
+        \Cx\Modules\Downloads\Controller\Category $child,
+        array &$names
+    ) {
+        if (empty($child->getParentId())) {
+            $names[] = $child->getName();
```
Comment (daeppen) [inline modules/Downloads/Controller/DownloadsManager.class.php]: should that not be done in any case? if so, then the statement on line [2528](https://github.com/Cloudrexx/cloudrexx/pull/132/files#diff-815346cc383dbccb8865f78c6680bba2R2528) is obsolete

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 6]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+            $title = $category['long'];
+            $text = $category['short'];
+
+            if ($category['short'] !== $category['long']) {
+                $text .= '...';
+            }
+
+        } else {
+            foreach ($category as $middle) {
+                $title .= '...' .$middle . "\n";
```
Comment (daeppen) [inline modules/Downloads/Controller/DownloadsManager.class.php]: pls remove the preceeding dots `'...'`

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 7]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+
+            $content =
+                sprintf(
+                    $_ARRAYLANG['TXT_DOWNLOADS_CATEGORIES_OF_CATEGORY'],
+                    '&bdquo;'.
+                    $this->getCategoryElement($categories['root']) .
+                    $this->getCategoryElement($categories['middle']) .
+                    $this->getCategoryElement($categories['previous']) .
+                    $this->getCategoryElement($categories['current']) .
+                    '&ldquo;'
```
Comment (daeppen) [inline modules/Downloads/Controller/DownloadsManager.class.php]: pls remove the leading and appending quotes (`&bdquo;` / `&ldquo;`) as they look displaced now with the new output format

File: modules/Downloads/Controller/DownloadsManager.class.php [hunk 8]
```diff
@@ -2401,10 +2401,136 @@ private function parseCategories($objCategory, $downloadOrderBy, $downloadOrderD
...
+
+            $categories = array(
+                'root' => array(),
+                'middle' => array(),
+                'previous' => array(),
+                'current' => array(),
+            );
+
+            foreach($parentNames as $position => $parentName) {
+                if ($position == 0) {
```
Comment (daeppen) [inline modules/Downloads/Controller/DownloadsManager.class.php]: pls change this `if-else-if`-construct into a `switch` statement

File: modules/Downloads/View/Template/Backend/module_downloads_overview.html [hunk 9]
```diff
@@ -38,6 +38,15 @@
 /* ]]> */
 </script>
 
+<style>
+    #downloads_category_form span:after {
+        content: ' / ';
```
[the remark this answers is not in the document]
Comment (Samangie) [reply modules/Downloads/View/Template/Backend/module_downloads_overview.html]: Done

File: modules/Downloads/View/Template/Backend/module_downloads_categories.html [hunk 10]
```diff
@@ -64,6 +64,15 @@
 // ]]>
 </script>
 
+<style>
+    #downloads_category_form span:after {
+        content: ' / ';
```
[the remark this answers is not in the document]
Comment (Samangie) [reply modules/Downloads/View/Template/Backend/module_downloads_categories.html]: Done

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 4 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: Done

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 5 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: Done

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 6 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: Done

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 7 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: Done

File: modules/Downloads/Controller/DownloadsManager.class.php (hunk 8 above)
Comment (Samangie) [reply modules/Downloads/Controller/DownloadsManager.class.php]: Done

**Pull request merged**